File: task2/stack/stack.cpp
#include "stack.h"

/*******顺序栈方法*******/
//输入参数：申请的顺序栈空间大小
template<class ElemType>
SqStack <ElemType> :: SqStack(int m)
{
	m_top = 0;
	m_base = new ELemType[m];
	printf("栈空间申请！");//Debug
	m_size = m;
}

template<class ElemType>
SqStack <ElemType> :: ~SqStack()
{
	if(m_base != NULL)
	{
		delete[] m_base;
		printf("栈空间释放！");//Debug
	}
}

template<class ElemType>
void SqStack <ElemType> :: Clear()
{
	m_top = 0;
}

template<class ElemType>
bool SqStack <ElemType> :: IsEmpty() const
{
	return m_top == 0;

}

template<class ElemType>
int SqStack <ElemType> :: Length() const
{
	return m_top;
}


template<class ElemType>
ElemType & SqStack <ElemType> :: Top() const
{
	if(m_top>0)
	{
		printf("取栈顶成功！");//Debug
		return m_base[m_top-1];
	}
	else 
		return NULL;
}

/*入栈*/
/*需先判断栈是否已经满了*/
template<class ElemType>
void SqStack <ElemType> :: Push(const ElemType &e)
{
	if(m_top>=m_size)          //如果栈空间不足，则申请新空间
	{
		ElemType *newbase;
		newbase = new ElemType[m_size + 10]; //新空间比原先大10
		for(int i = 0; i < m_top; i++)       //把旧栈的数据复制到新站  
		{
			*(newbase + i) = *(m_base + i);
		} 
		delete[] m_base;                      //释放原空间 
		m_base = newbase;                     //栈空间转移完毕
		m_size += 10;                         //更新新栈空间大小
	}

	m_base[m_top++] = e;                      //入栈

}

/*出栈*/
template<class ElemType>
void SqStack <ElemType> :: Pop()
{
	if(m_top > 0)
	{
		m_top--;
	}

}

/*******链栈方法*******/
template <class ElemType>
LinkStack <ElemType>::LinkStack()
{
	top = NULL;

}

template <class ElemType>
LinkStack <ElemType>::~LinkStack()
{}

template <class ElemType>
void LinkStack <ElemType>::Clear()
{}

template <class ElemType>
bool LinkStack <ElemType>::IsEmpty() const
{}

template <class ElemType>
LinkStack <ElemType>::Length() const
{}

template <class ElemType>
LinkStack <ElemType>::Top() const
{}

template <class ElemType>
LinkStack <ElemType>::Push(ElemType &e) 
{}

template <class ElemType>
LinkStack <ElemType>::Pop() 
{}

File: task2/stack/stack.py

##顺序栈
class Arrstack:
	def __init__(self):
		self.stack = []
	def __str__(self):
		return self.stack.__str__()
	def pop(self):
		return self.stack.pop()
	def push(self, elem):
		self.stack.append(elem)
	@property
	def size(self):
		return len(self.stack)
	@property
	def is_empty(self):
		return len(self.stack) == 0
	def display(self):
		print(self.stack)

##恋栈节点
class Node:
	def __init__(self, value = None, next = None):
		self.value = value
		self.next = None
	##使得节点可以直接打印
	def __str__(self):
		return str(self.value)
##恋栈
class Linkstack:
	_top = None
	def __init__(self):
		pass
	##这个方法作为类属性一样使用，不用加括号
	@property
	def is_empty(self):
		return self._top is None

	def push(self, value):
		node = self._top
		self._top = Node(value)
		self._top.next = node
		return True

	def pop(self):
		if self.is_empty:
			return None
		node = self._top
		self._top = node.next
		return node 
	##生成器，生成整个栈的迭代器
	def _traversal(self):
		node = self._top
		##注如果栈为空生成空迭代器
		while node :
			yield node 
			node = node.next
	def display(self):
		print('\n'.join(map(lambda node: '|{:^7}|'.format(str(node) ), self._traversal() ) ) + '\n' + 7 * '-' )



if __name__ == '__main__':
	# stack_test = Linkstack()
	# # for i in range(10):
	# # 	stack_test.push(i)
	# print(stack_test._traversal())
	# it = stack_test._traversal()
	# next(it)
	# stack_test.display()
	stack_t =	Arrstack()
	for i in range(10):
		stack_t.push(i)
	print(stack_t)
	print(stack_t.size)
	print(stack_t.is_empty)
	for i in range(10):
		x = stack_t.pop()
		print(stack_t)
		#print('pop %d'%x )
	print(stack_t.size)
	print(stack_t.is_empty)